### PR TITLE
Resize defect vector when changing multigrid levels

### PR DESCRIPTION
**Pull request by masterleinad**
Furthermore, I don't see any particular reason why the `Multigrid` constructor that let the user specify minimal and maximal level is considered to be problematic.

### CODE REVIEW

**Comment by davydden** [comment]
should not one resize `solution`, `t`, `defect2` as well? The former two are initialised in constructor but are not altered in `reinit` at the moment.

**Comment by davydden** [comment]
never mind, i just realised that they are resized in `cycle()`.

**Comment by davydden** [comment]
i supose `copy_to_mg()` is fine when defect is resized to custom minlevel?

**Comment by masterleinad** [comment]
Yes, `min_level!=0` is tested in `multigrid/step-16-03.cc` for example.

**Comment by davydden** [comment]
LGTM

File: include/deal.II/multigrid/multigrid.templates.h [hunk 1]
```diff
@@ -69,9 +72,7 @@ template <typename VectorType>
 void
 Multigrid<VectorType>::set_maxlevel (const unsigned int l)
 {
-  Assert (l <= maxlevel, ExcIndexRange(l,minlevel,maxlevel+1));
-  Assert (l >= minlevel, ExcIndexRange(l,minlevel,maxlevel+1));
-  maxlevel = l;
```
Comment (tjhei) [inline include/deal.II/multigrid/multigrid.templates.h]: Why are you not setting ``maxlevel``?

Comment (davydden) [inline include/deal.II/multigrid/multigrid.templates.h]: it's in `reinit` call below.

**Comment by tjhei** [comment]
> Furthermore, I don't see any particular reason why the Multigrid constructor that let the user specify minimal and maximal level is considered to be problematic.

minlevel!=0 didn't use to work for the longest time, that is why. I am also not sure if anyone used a non-standard maxlevel before.

File: include/deal.II/multigrid/multigrid.h [hunk 2]
```diff
@@ -99,9 +99,8 @@ class Multigrid : public Subscriptor
             Cycle                              cycle = v_cycle);
 
   /**
-   * Experimental constructor for cases in which no DoFHandler is available.
-   *
-   * @warning Not intended for general use.
+   * Constructor. Same as above but you can determine the multigrid
+   * levels to use yourself.
```
Comment (tjhei) [inline include/deal.II/multigrid/multigrid.h]: One problem I have with this method is that we don't have the ``DoFHandler`` here so we can not check if the maxlevel the user specifies is valid.

File: include/deal.II/multigrid/multigrid.templates.h (hunk 1 above)
Comment (tjhei) [inline include/deal.II/multigrid/multigrid.templates.h]: ah, got it.

File: include/deal.II/multigrid/multigrid.h (hunk 2 above)
Comment (davydden) [inline include/deal.II/multigrid/multigrid.h]: why would one need this constructor at all? What's the usage case?
It's possible to override `minlevel` and `maxlevel` using corresponding methods while using the constructor with dof_handler.

Comment (tjhei) [inline include/deal.II/multigrid/multigrid.h]: Agreed. You could add min/maxlevel with default arguments to the contructor above, too...

Comment (davydden) [inline include/deal.II/multigrid/multigrid.h]: so maybe just get rid of this constructor altogether to avoid confusion for users?

Comment (masterleinad) [inline include/deal.II/multigrid/multigrid.h]: The `Multigrid` class doesn't use a `DoFHandler` apart from setting default values for the levels we want to work on. Therefore, I like having a constructor that can be used without providing a `DoFHandler`. Unfortunately, none of the other objects that are used in the constructor provides us with the minimum and maximum level they have. I would be in favor of changing this, i.e. get this information from ` MGMatrixBase`.

I don't care too much what we do in the other case. If you are in favor of having the possibility to specify `minlevel` and `maxlevel` in the first constructor as well, I am fine with including it.

Comment (davydden) [inline include/deal.II/multigrid/multigrid.h]: on another hand, when doing GMG in deal.II you are bound to have a dof handler object around and mesh and alike. So do we really win anything with this constructor?

Comment (davydden) [inline include/deal.II/multigrid/multigrid.h]: @kronbichler what's your opinion on this constructor?

**Comment by tjhei** [comment]
How about changing the interface of the constructor to match the arguments the other constructor has (only add min/max levels)? I think that makes the most sense.

**Comment by kronbichler** [comment]
Hidden question by @davydden 
> @kronbichler what's your opinion on this constructor [with `minlevel` and `maxlevel` as first two arguments and without DoFHandler]?

I'm also more in favor of providing one single constructor where `minlevel` and `maxlevel` are optional. I think this can be merged as is in my opinion, given that we have the more general PR #2534 open that discusses the general design of a better MG infrastructure (the purpose of that issue was primarily performance on really large computations with >100k processors but we should definitely also think about the design).

More generally, I am also in favor of extending the capabilities of `mg::Matrix` such that it gains knowledge about the minimum and maximum level.

**Comment by masterleinad** [comment]
mg::Matrix gains knowledge about minimum and maximum level and can be used instead of the DoFHandler to determine valid level bounds. This makes the constructor that takes a DoFHandler redundant.

**Comment by davydden** [comment]
@masterleinad 

I suggested in https://github.com/dealii/dealii/issues/2534#issuecomment-212887225 , that we should remove `mg::Matrix` as it does not really bring anything but, IMHO, confusion with `MGLevelObject<matrix_t>` which is used to store matrices. 

Can you explain in what circumstances you would like to use this constructor without `DoFHandler` around? In other words, why is it important not to have this constructor?

**Comment by davydden** [comment]
@masterleinad 
at the time of discussion (Apr 21) @kronbichler was (in my interpretation) of a similar opinion that we may not need this class, see https://github.com/dealii/dealii/issues/2534#issuecomment-213031932 .

**Comment by masterleinad** [comment]
The whole point of the last commit was to get rid off the constructor with `DoFHandler` as the `DoFHandler` is not used apart from checking level bounds. The matrix object is the one that should know about which levels are available:

@kronbichler :
> More generally, I am also in favor of extending the capabilities of mg::Matrix such that it gains knowledge 
> about the minimum and maximum level.

In my opinion the multigrid implementation should not necessarily need a `DoFHandler` object. In principle, 
`MGMatrixBase`, `MGCoarseGridBase`, `MGTransferBase`, `MGSmootherBase` can just use suitable matrices.

I am fine with using `MGLevelObject<matrix_t>` instead of `MGMatrixBase`. From the discussion in #3092, 

> The point is just that a somewhat unified interface to the multigrid algorithms helps. If we go for templates in one place, we might just as well go with templates for both the smoother, the matrix, and the coarse grid solver.

I got the impression that we wanted to have a similar design for `MGMatrixBase`, `MGCoarseGridBase`, `MGTransferBase`, `MGSmootherBase` so I didn't change it immediately and retain backward-compatibility.

**Comment by kronbichler** [comment]
I added a discussion in #2534 about how the Multigrid class should ideally look like. We should agree on those on a basic level.

Unless there is a new consensus on that issue during the next few days, I would propose to stay with the current solution as of now and merge this PR.

**Comment by kronbichler** [comment]
Given that we agreed upon improving the current design in #2534 and this definitely contributes to that I will merge this.

**Pull request merged**